numpy/core/numeric.py

```diff
@@ -604,6 +607,53 @@ def asfortranarray(a, dtype=None):
     """
     return array(a, dtype, copy=False, order='F', ndmin=1)
 
+def asanycontiguousarray(a, dtype=None):
+    """
+    Return a contiguous array in memory (either C or Fortran order).
+    If the input is already C- or Fortran-contiguous, its memory
+    order is preserved.
+
+    Parameters
+    ----------
+    a : array_like
+        Input array.
+    dtype : str or dtype object, optional
+        Data-type of returned array.
+
+    Returns
+    -------
+    out : ndarray
+        C- or Fortran-contiguous array of same shape and content as `a`,
+        with type `dtype` if specified.  A copy is made only if needed.
+
+    See Also
+    --------
+    ascontiguousarray : Convert input to a contiguous (C order) array.
+    asfortranarray : Convert input to an ndarray with column-major
+                     memory order.
+    require : Return an ndarray that satisfies requirements.
+    ndarray.flags : Information about the memory layout of the array.
+
+    Examples
+    --------
+    >>> x = np.arange(6).reshape(2,3)
+    >>> np.asanycontiguousarray(x, dtype=np.float32)
+    array([[ 0.,  1.,  2.],
+           [ 3.,  4.,  5.]], dtype=float32)
+    >>> x.flags['C_CONTIGUOUS']
+    True
+
+    """
+    try:
+        is_f_contiguous = a.flags.f_contiguous
+    except AttributeError:
+        # Not a ndarray
+        is_f_contiguous = False
+    if is_f_contiguous:
+        return array(a, dtype, copy=False, order='F', ndmin=1)
+    else:
+        return array(a, dtype, copy=False, order='C', ndmin=1)
```

Comment:
Not sure, I like the idea too much, but that is not a deal breaker (also the `any` has a bit of a name clash with `asanyarray`). But I think you can simplify this to a single line function:
```
return array(a, dtype, copy=False, order="A", ndmin=1)
```
(not that I like the `ndmin=1` either, what is so bad about 0-D arrays? But I guess the ascontiguousarray functions do that too). The whole f_contiguous stuff should be handled by "A" is my guess.